lib/class-wp-rest-annotations-controller.php

```diff
@@ -0,0 +1,1097 @@
+		$response = $this->prepare_item_for_response( $comment, $request );
+		$response = rest_ensure_response( $response );
+
+		$response->set_status( 201 );
+
+		// phpcs:ignore PHPCompatibility.PHP.NewKeywords.t_namespaceFound — mistaken as 'namespace' keyword.
+		$response->header( 'Location', rest_url( sprintf( '%s/%s/%d', $this->namespace, $this->rest_base, $comment_id ) ) ); // @codingStandardsIgnoreLine
+
+		return $response;
+	}
+
+	/**
+	 * Checks if request has access to create an item.
+	 *
+	 * @since [version]
+	 *
+	 * @param  WP_REST_Request $request Full REST API request details.
+	 *
+	 * @return bool|WP_Error            True if request has access to create,
+	 *                                  {@see WP_Error} otherwise.
+	 *
+	 * @see WP_REST_Comments_Controller::create_item_permissions_check()
+	 */
+	public function create_item_permissions_check( $request ) {
+		$post_id        = $request['post'];
+		$type           = $request['type'];
+		$status         = $request['status'];
+		$parent_id      = $request['parent'];
+		$default_params = $request->get_default_params();
+
+		if ( ! $post_id ) {
+			return new WP_Error( 'rest_missing_annotation_post', __( 'Missing post.', 'gutenberg' ), array(
+				'status' => 400,
+			) );
+		}
+
+		if ( ! $type ) {
+			return new WP_Error( 'rest_invalid_annotation_param_type', __( 'Missing type.', 'gutenberg' ), array(
+				'status' => 400,
+			) );
+		}
+
+		if ( ! $status && '0' !== $status ) {
+			return new WP_Error( 'rest_invalid_annotation_param_status', __( 'Missing status.', 'gutenberg' ), array(
+				'status' => 400,
+			) );
+		}
+
+		if ( ! $this->check_read_post_permission( $post_id, $request ) ) {
+			return new WP_Error( 'rest_cannot_read_annotation_post', __( 'Sorry, you are not allowed as this user.', 'gutenberg' ), array(
+				'status' => rest_authorization_required_code(),
+			) );
+		}
+
+		if ( ! current_user_can( 'create_annotation', $post_id, $type ) ) {
+			return new WP_Error( 'rest_cannot_create_annotation', __( 'Sorry, you are not allowed as this user.', 'gutenberg' ), array(
+				'status' => rest_authorization_required_code(),
+			) );
+		}
+
+		if ( $parent_id && ! current_user_can( 'read_annotation', $parent_id ) ) {
+			return new WP_Error( 'rest_cannot_read_annotation_parent', __( 'Sorry, you are not allowed as this user.', 'gutenberg' ), array(
+				'status' => rest_authorization_required_code(),
+			) );
+		}
+
+		foreach ( array(
+			'id',
+		) as $param ) {
+			if ( isset( $request[ $param ] ) && ( ! isset( $default_params[ $param ] ) || $request[ $param ] !== $default_params[ $param ] ) ) {
+				return new WP_Error( 'rest_readonly_annotation_param_' . $param, __( 'Read-only param.', 'gutenberg' ), array(
+					'status' => 400,
+				) );
+			}
+		}
+
+		// @codeCoverageIgnoreStart
+		// Not possible to test at this time.
+		// Strengthens security in custom post types and on sites with permission mods.
+		if ( ! current_user_can( 'edit_annotations', $type ) ) {
```

Comment:
Great suggestion. I'll elaborate on this in the comments. The rationale for this is not great at the moment, as it is currently impossible to `create_annotations` if you can't `edit_annotations`.

I left this in only because it would prevent date/meta from being set in the case of a custom post type or plugin altering permissions in such a way that it allows annotation creation, but locks editing. Those two fields would then be restricted in such a case, which seems desirable in that scenario.

I'm going review this again though, because it _is_ very confusing and way out on the edge. I'd prefer to get rid of it in favor of a more straightforward approach.

Reply:
Hmm, but I would argue that you should be allowed to at least set meta on an annotation if you are not allowed to edit it. For example, in the scenario where we would persist annotations from Yoast SEO, we would like to save some extra data in meta. But then users who can only create annotation (not edit) would not be able to set this extra data.

Reply:
Good point. I have no problem loosening permissions here by removing `meta` from the list.

A scenario where a user can create, but not edit, is not one that would occur unless permissions were altered by a plugin. I will leave a portion of this in to protect the `date` and `date_gmt` fields in that scenario, but I'll remove `meta` as you point out.